Add Django ASGI support

adamantike opened the pull request:
# Description

Still a WIP, this diff adds `asgi` as an extra, and uses its methods if the current request is an `ASGIRequest`.

I still need to dig deeper in the current test suite, to find a way to duplicate the tests in
`instrumentation/opentelemetry-instrumentation-django/tests/test_middleware.py`, but using an [`AsyncClient`](https://docs.djangoproject.com/en/3.1/topics/testing/tools/#testing-asynchronous-code).

Fixes #165, #185, #280, #334.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [ ] Unit tests have been added
- [ ] Documentation has been updated

adamantike commented:
Marking this as ready for review. There are still a few things to investigate:

* Tests are passing, but triggering log error: `opentelemetry.context:__init__.py:146 Failed to detach context`.
* ASGI intrumentation's `set_status_code` method does not populate the `status_text` field. I think this should be tackled in a separate PR.
* Traced request attributes are not supported in this PR yet.

cc @lonewolf3739, because of the context provided in #334.

ocelotl commented:
@adamantike is this still WIP? If so, please mark it as draft.

ocelotl commented:
> @adamantike is this still WIP? If so, please mark it as draft.

Oh, I see, please edit the PR description instead.

ocelotl commented on instrumentation/opentelemetry-instrumentation-django/tests/test_middleware_asgi.py:
```diff
@@ -0,0 +1,276 @@
+        span = span_list[0]
+        self.assertEqual(span.name, "HTTP GET")
+
+    @pytest.mark.skip(
```
I think it would be better not to have this test at all instead of skipping it every time and add it after traced request attributes are supported. That way it won't show up constantly as being skipped in the test runs.

adamantike replied:
I have added support for traced request attributes, so unskipped the test instead :)

adamantike commented on instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:
```diff
@@ -113,7 +113,7 @@ def get_host_port_url_tuple(scope):
     server = scope.get("server") or ["0.0.0.0", 80]
     port = server[1]
-    server_host = server[0] + (":" + str(port) if port != 80 else "")
+    server_host = server[0] + (":" + str(port) if str(port) != "80" else "")
```
I found this discrepancy in [Django's `AsyncRequestFactory`](https://github.com/django/django/blob/a948d9df394aafded78d72b1daa785a0abfeab48/django/test/client.py#L500) while testing, as the base scope is created with a port `'80'`, instead of using an integer.

I think that changing this check could avoid any issues with other potential callers using a string instead of an integer.

adamantike commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -182,7 +230,10 @@ def process_response(self, request, response):
+            if is_asgi_request:
+                detach(request.META.get(self._environ_token))
+            else:
+                detach(request.environ.get(self._environ_token))
```
Consider that `request.environ` is only [defined in `WSGIRequest`](https://github.com/django/django/blob/a948d9df394aafded78d72b1daa785a0abfeab48/django/core/handlers/wsgi.py#L70), but not set in `ASGIRequest`, so moving this line out of the `else` block would trigger an `AttributeError` if we're processing an asynchronous request and try to execute `detach(request.environ.get(self._environ_token))`.

lzchen replied:
How come we can't `detach(request.META.get(self._environ_token))` for both?

adamantike commented on instrumentation/opentelemetry-instrumentation-django/tests/test_middleware_asgi.py:
```diff
@@ -0,0 +1,276 @@
+        super().tearDownClass()
+        conf.settings = conf.LazySettings()
+
+    @classmethod
+    def _add_databases_failures(cls):
+        # Disable databases.
+        pass
+
+    @classmethod
+    def _remove_databases_failures(cls):
+        # Disable databases.
+        pass
```
The purpose of these methods was to override Django's provided ones, which try to validate and database connections: https://github.com/django/django/blob/a948d9df394aafded78d72b1daa785a0abfeab48/django/test/testcases.py#L202-L226

I'll recheck if they are still needed, or if we can delete them, though.

adamantike replied:
Found the methods unnecessary after avoiding inheriting from `SimpleTestCase` class from Django, and correctly setting up Django settings in `setUpClass`!

adamantike commented:
I have rebased, fixed some comments, and added support for traced request attributes! The only issue I still find is this one:

* Tests are passing, but triggering log error: opentelemetry.context:__init__.py:146 Failed to detach context.

However, it could be related to how Pytest handles ContextVars and contexts related to attached tokens, based on https://github.com/pytest-dev/pytest-asyncio/issues/127

adamantike commented:
Please do not merge yet. I just found that the `Failed to detach context` log messages also appear ina real Django project, when manually installing this branch, for a project using Gunicorn with Uvicorn workers.

adamantike commented:
I have filed https://code.djangoproject.com/ticket/32815, to determine if the bug is in Django's codebase instead.

adamantike commented:
Filed https://github.com/open-telemetry/opentelemetry-python-contrib/pull/533, which circumvents the `Failed to detach context` error by simplifying how the Django middleware is wrapped with `sync_to_async`. With that change applied, the issue doesn't occur anymore!

adamantike commented:
Rebased now that #533 was merged. All the `Failed to detach context` errors have been fixed, and this is now ready for another review!

ocelotl commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -69,6 +70,26 @@ def __call__(self, request):
 
 
+try:
+    from django.core.handlers.asgi import ASGIRequest
```
Better to check for version instead, an import error may not be caused necessarily by this component being missing.

adamantike replied:
I agree, also maintains consistency with the version check we are now doing to decide the Middleware style being used.

ocelotl commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -92,6 +113,10 @@ def __call__(self, request):
 
 
+def _is_asgi_request(request: HttpRequest) -> bool:
+    return bool(ASGIRequest and isinstance(request, ASGIRequest))
```
```suggestion
    return ASGIRequest is not None and isinstance(request, ASGIRequest)
```

ocelotl commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -156,12 +192,36 @@ def process_request(self, request):
+                        **{
+                            **request.__dict__,
+                            **{
+                                name.decode("latin1"): value.decode("latin1")
```
Just curious, why `"latin1"`?

adamantike replied:
Here, I'm using the same logic Django uses to decode the request scope in ASGI mode, which uses `latin1`. This is the relevant upstream code: https://github.com/django/django/blob/7bb0605a8c4adb4247a179276036f138ae859ac9/django/core/handlers/asgi.py#L76-L90

ocelotl reviewed (changes requested):
Looking good! Just minor changes requested.

adamantike commented:
@ocelotl, thanks for the feedback! Pushed the requested changes.

owais commented on instrumentation/opentelemetry-instrumentation-django/setup.cfg:
```diff
@@ -45,6 +45,8 @@ install_requires =
 
 [options.extras_require]
+asgi =
+    opentelemetry-instrumentation-asgi == 0.23.dev0
```
This should be documented so people using ASGI with Django realize they need to install these extra deps to make it work. 

How do ASGI deps work with Django? Do regular Django projects need to install additional packages to make ASGI work or does Django now ship with (or depend on) bundled ASGI module.  If some level of ASGI support is always present in Django now, may be we can add this as a main dependency instead?

adamantike replied:
This is a good discussion topic. Django doesn't require any additional/extra dependencies for ASGI to work (aside from running it using Uvicorn/Daphne), and bundles `asgiref` since the async introduction in Django 3.0: https://github.com/django/django/commit/a415ce70bef6d91036b00dd2c8544aed7aeeaaed#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7

This dependency doesn't include other heavy dependencies that aren't already required by `opentelemetry-instrumentation-wsgi`. However, we also have to consider that ASGI is only present since Django 3.x, while this project tries to maintain support for older Django versions.

I think the only missing change on this PR, if we go with adding this to `install_requires`, is to make `_is_asgi_supported` also depend on `DJANGO_3_0` being `True`, as we won't receive any `ImportError` anymore when importing ASGI dependencies.

ocelotl commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -92,6 +113,10 @@ def __call__(self, request):
 
 
+def _is_asgi_request(request: HttpRequest) -> bool:
+    return ASGIRequest is not None and isinstance(request, ASGIRequest)
```
`ASGIRequest` can be `None`, in that case this suggested approach will cause an error:

```python
In [1]: isinstance(1, None)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-ce5efb9b0eda> in <module>
----> 1 isinstance(1, None)

TypeError: isinstance() arg 2 must be a type or tuple of types
```

Also, if checking that a certain object is not `None` the `is not None` form [should be used](https://www.python.org/dev/peps/pep-0008/#programming-recommendations).

owais replied:
I think there is also an is_asgi_supported variable being set above. we could also use that and then run instance check

adamantike replied:
I don't think we should mix the concepts of `is_asgi_supported` (based on required Python packages being installed) from `_is_asgi_request` (based on the current request being handled), considering that the Django middleware can handle both WSGI and ASGI requests, and most of the conditionals are `if is_asgi_request and not _is_asgi_supported` to return early for invalid tracing scenarios.

owais commented:
@adamantike Hi, thanks for taking this up. Are you still planning to finish this?

adamantike commented:
@owais, I'll be tackling your comments this week, so we can move forward! Thanks for all the feedback

adamantike commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -141,12 +166,23 @@ def process_request(self, request):
 
+        is_asgi_request = _is_asgi_request(request)
+        if is_asgi_request and not _is_asgi_supported:
+            return
```
> Can a django service/middleware handle both WSGI and ASGI requests at the same time?
Yes, it can handle both, so we need to check which kind of request we're receiving on each call.

> At least is_asgi_supported can be called before _is_asgi_request and that way we can avoid running the type check for non-ASGI requests.
As we're using `is_asgi_request` multiple times on this method, the code will need to be a bit more convoluted if we want to avoid the type check when we know ASGI is not supported. I will change the code accordingly, and we can decide if it's worth to keep it that way.

adamantike commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -156,12 +192,36 @@ def process_request(self, request):
+        if is_asgi_request:
+            attributes = collect_request_attributes(request.scope)
+        else:
+            attributes = collect_request_attributes(request_meta)
```
Good idea, applied!

adamantike commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -156,12 +192,36 @@ def process_request(self, request):
+                # ASGI requests include extra attributes in request.scope.headers. For this reason,
+                # we need to build an object with the union of `request` and `request.scope.headers`
+                # contents, for the extract_attributes_from_object function to be able to retrieve
+                # attributes from it.
```
Would it make more sense, to keep using the `extract_attributes_from_object` util, to go with a composition approach?

```python
attributes = extract_attributes_from_object(
    request, self._traced_request_attrs, attributes
)
if is_asgi_request:
    # ASGI requests include extra attributes in request.scope.headers.
    attributes = extract_attributes_from_object(
        types.SimpleNamespace(**{
            name.decode("latin1"): value.decode("latin1")
            for name, value in request.scope.get("headers", [])
        }),
        self._traced_request_attrs,
        attributes,
    )
```

adamantike commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -208,15 +268,24 @@ def process_response(self, request, response):
-                "{} {}".format(response.status_code, response.reason_phrase),
-                response,
-            )
+            if is_asgi_request:
```
As the functions are right now, I don't see any needed changes for both WSGI and ASGI to make the same changes to the current span:
* WSGI: https://github.com/open-telemetry/opentelemetry-python-contrib/blob/e12917401f5ca6af8737378a99b00da2b1987d5c/instrumentation/opentelemetry-instrumentation-wsgi/src/opentelemetry/instrumentation/wsgi/__init__.py#L188-L208
* ASGI: https://github.com/open-telemetry/opentelemetry-python-contrib/blob/e12917401f5ca6af8737378a99b00da2b1987d5c/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L125-L140

adamantike commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -208,15 +268,24 @@ def process_response(self, request, response):
-                response,
-            )
+            if is_asgi_request:
+                set_status_code(span, response.status_code)
```
I also think this is already fixed in the current `set_status_code` implementation.

adamantike commented on instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware.py:
```diff
@@ -68,6 +70,25 @@ def __call__(self, request):
+else:
+    ASGIRequest = None
+
+try:
```
Applied, thanks!

adamantike commented:
@owais @ocelotl, sorry for the delay. This is now ready for another review!

owais commented:
@adamantike sorry for the delay. can you please resolve the conflicts and rebase/merge with main? We can try to include it in the release this week.

adamantike commented:
I agree, though by that same approach, a future breaking change should be introduced for `wsgi` to also become an extra, so users exclusively having ASGI don't get WSGI dependencies.

This is where https://github.com/pypa/setuptools/issues/1139 would come useful if implemented. If no extras are specified, you get both WSGI and ASGI. If you want to cut down on dependencies, you need to specify which flow your app is going to use.

owais commented:
Oh I've been waiting on pypa/setuptools#1139 for longer than I remember. It's such an obvious feature to have :)
